src/en/developer-layers-interfaces2.md

```diff
@@ -0,0 +1,228 @@
+eventually agree or be set by a single leader.
+
+```python
+class MyRelationClient(RelationBase):
+  scope = Scopes.SERVICE
+```
+
+#### UNIT
+
+Each connected unit for this relation will have its own conversation. This is
+the default scope. Each unit’s data will be retrieved individually, but note
+that due to how Juju works, the same data is still broadcast to all units of a
+single service.
+
+```python
+class MyRelationClient(RelationBase):
+  scope = scopes.UNIT
+```
+
+## Writing an interface-layer
+
+Begin by making an interface repository if you don't currently have one.
+
+```bash
+mkdir -p $JUJU_REPOSITORY/interfaces
+export INTERFACE_PATH=$JUJU_REPOSITORY/interfaces
+```
+
+The export of `INTERFACE_PATH` is an environment variable which tells the
+`charm build` process where to scan for local interfaces not found in the
+[layer registry](http://interfaces.juju.solutions).
+
+With our interface repository created, we can now create our new interface.
+
+Start by creating the directory to warehouse your interface
+
+```Bash
+mkdir -p $INTERFACE_PATH/http
+```
+
+And declare the interface's metadata in `interface.yaml`
+
+```yaml
+name: http
+summary: Basic HTTP interface
+version: 1
+repo: https://git.launchpad.net/~bcsaller/charms/+source/http
+```
+
+#### Writing the provides side
+
+We're now ready to implement the provider interface in `provides.py`.
+
+```python
+from charmhelpers.core import hookenv
+from charms.reactive import hook
+from charms.reactive import RelationBase
+from charms.reactive import scopes
+
+
+class HttpProvides(RelationBase):
+    # Every unit connecting will get the same information
+    scope = scopes.GLOBAL
+
+    # Use some template magic to declare our relation(s)
+    @hook('{provides:http}-relation-{joined,changed}')
+    def changed(self):
+        # Signify that the relationship is now available to our principal layer(s)
+        self.set_state('{relation_name}.available')
+
+    @hook('{provides:http}-relation-{broken,departed}')
+    def broken(self):
+        # Remove the state that our relationship is now available to our principal layer(s)
+        self.remove_state('{relation_name}.available')
+
+    # call this method when passed into methods decorated with
+    # @when('{relation}.available')
+    # to configure the relation data
+    def configure(self, port):
+        relation_info = {
+            'hostname': hookenv.unit_get('private-address'),
+            'port': port,
+        }
+        self.set_remote(**relation_info)
+```
+
+#### Implementing the provides side
+
+With our provider interface written, lets take a look at how we might implement
+this in a charm.
+
+In our metadata we define a website relation implementing the interface:
+```yaml
+provides:
+  website:
+    interface: http
+```
+
+And in the reactive class, we will receive the RelationBase object, and invoke
+the configure method:
+
+```python
+@when('website.available')
+def configure_website(website):
```

Comment:
It always confuses me where the argument (website) comes from. You have explained it a few times, but it is worth calling out "website" in this example.